### PR TITLE
Handle multiple upstreams in ingress-controller

### PR DESCRIPTION
**Pull request by ChrsMark**
## What does this PR do?
This PR tunes `ingress-controller`'s pipeline so as to handle multiple upstreams. We put the multiple lengths, times and status_codes in `upstream.response.length_list`, `upstream.response.time_list` and `upstream.response.status_code_list` fields while we sum time under single field `upstream.response.time` and we only store last response length in `upstream.response.length` (with this we also preserve the existing dashboards https://github.com/elastic/beats/blob/54ea2842471a33c8f23833393081ab65b3c54a04/filebeat/module/nginx/_meta/kibana/7/dashboard/Filebeat-nginx-ingress-overview.json#L1007).

In addition, we preserve the multiple upstreams at `upstream_ip_list` and we only store the last `upstream.ip` and `upstream.port` as well as the last `status_code`.

Sample log:
`192.168.64.14 - - [07/Feb/2020:12:02:42 +0000] "GET /v2/some HTTP/1.1" 200 61 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.14; rv:72.0) Gecko/20100101 Firefox/72.0" 348 0.001 [default-web2-8080] [] 172.17.0.6:8080, 172.17.0.7:8080 61, 100 0.100, 0.004 200, 203 835136ae24486dbb4156dcbe21f5d402
`
Sample event:
```
{
        "nginx.ingress_controller.http.request.id": "835136ae24486dbb4156dcbe21f5d402",
        "nginx.ingress_controller.http.request.length": 348,
        "nginx.ingress_controller.http.request.time": 0.001,
        "nginx.ingress_controller.remote_ip_list": [
            "192.168.64.14"
        ],
        "nginx.ingress_controller.upstream.alternative_name": "",
        "nginx.ingress_controller.upstream.ip": "172.17.0.7",
        "nginx.ingress_controller.upstream.name": "default-web2-8080",
        "nginx.ingress_controller.upstream.port": 8080,
        "nginx.ingress_controller.upstream.response.length": 100,
        "nginx.ingress_controller.upstream.response.length_list": [
            "61",
            "100"
        ],
        "nginx.ingress_controller.upstream.response.status_code": 203,
        "nginx.ingress_controller.upstream.response.status_code_list": [
            "200",
            "203"
        ],
        "nginx.ingress_controller.upstream.response.time": 0.104,
        "nginx.ingress_controller.upstream.response.time_list": [
            "0.100",
            "0.004"
        ],
        "nginx.ingress_controller.upstream_address_list": [
            "172.17.0.6:8080",
            "172.17.0.7:8080"
        ],
}
```

## Why is it important?
So as to handle cases where nginx-ingress retries multiple upstreams.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Related issues

- Closes https://github.com/elastic/beats/issues/20813

### CODE REVIEW

File: filebeat/module/nginx/ingress_controller/ingest/pipeline.yml [hunk 1]
```diff
@@ -41,6 +61,81 @@ processors:
...
+    source: >-
+      try {
+        if (ctx.nginx.ingress_controller.upstream.response.length_list.length == null) {
+          return;
+        }
+        int length = 0;
+        for (def item : ctx.nginx.ingress_controller.upstream.response.length_list) {
+          length = length + Integer.parseInt(item);
+        }
+        ctx.nginx.ingress_controller.upstream.response.length = length;
```
Comment (kaiyan-sheng) [inline filebeat/module/nginx/ingress_controller/ingest/pipeline.yml:76]: Maybe change the name to indicate it's a total value?
```suggestion
        ctx.nginx.ingress_controller.upstream.response.length_total = length;
```

Comment (jsoriano) [reply filebeat/module/nginx/ingress_controller/ingest/pipeline.yml]: Take into account that this is a breaking change as it changes an existing field. I would prefer to keep the original field.

File: filebeat/module/nginx/ingress_controller/_meta/fields.yml [hunk 2]
```diff
@@ -11,6 +11,26 @@
         Real source IP is restored to `source.ip`.
 
     # ingress-controller specific fields
+    - name: upstream_ip_list
+      type: array
```
Comment (jsoriano) [inline filebeat/module/nginx/ingress_controller/_meta/fields.yml:15]: I think that `type: array` only doesn't generate any mapping. This should be an array of keywords, or just keyword. Elasticsearch doesn't have a data type for arrays, but any field can have multiple values, see [docs about arrays](https://www.elastic.co/guide/en/elasticsearch/reference/7.9/array.html)).

We could use just keyword:
```suggestion
      type: keyword
```

Or if we still want to indicate that this is intended to store multiple values:
```suggestion
      type: array
      object_type: keyword
```

Same for the other lists.

File: filebeat/module/nginx/ingress_controller/ingest/pipeline.yml [hunk 3]
```diff
@@ -12,14 +12,18 @@ processors:
...
-      (%{NUMBER:nginx.ingress_controller.upstream.response.length:long}|-) (%{NUMBER:nginx.ingress_controller.upstream.response.time:double}|-)
-      (%{NUMBER:nginx.ingress_controller.upstream.response.status_code:long}|-) %{GREEDYDATA:nginx.ingress_controller.http.request.id}
+      \[%{DATA:nginx.ingress_controller.upstream.alternative_name}\] (%{UPSTREAM_ADDRESS_LIST:nginx.ingress_controller.upstream_ip_list}|-)
+      (%{UPSTREAM_RESPONSE_LENGTH_LIST:nginx.ingress_controller.upstream.response.length_list}|-) (%{UPSTREAM_RESPONSE_TIME_LIST:nginx.ingress_controller.upstream.response.time_list}|-)
+      (%{UPSTREAM_RESPONSE_STATUS_CODE_LIST:nginx.ingress_controller.upstream.response.status_code_list}|-) %{GREEDYDATA:nginx.ingress_controller.http.request.id}
     pattern_definitions:
       NGINX_HOST: (?:%{IP:destination.ip}|%{NGINX_NOTSEPARATOR:destination.domain})(:%{NUMBER:destination.port})?
       NGINX_NOTSEPARATOR: "[^\t ,:]+"
       NGINX_ADDRESS_LIST: (?:%{IP}|%{WORD})("?,?\s*(?:%{IP}|%{WORD}))*
       UPSTREAM_ADDRESS: '%{IP:nginx.ingress_controller.upstream.ip}(:%{NUMBER:nginx.ingress_controller.upstream.port})?'
```
Comment (jsoriano) [inline filebeat/module/nginx/ingress_controller/ingest/pipeline.yml:22]: Not used now.

File: filebeat/module/nginx/ingress_controller/test/test.log-expected.json [hunk 4]
```diff
@@ -28,18 +28,30 @@
...
         "nginx.ingress_controller.upstream.response.status_code": 200,
-        "nginx.ingress_controller.upstream.response.time": 0.0,
+        "nginx.ingress_controller.upstream.response.status_code_list": [
+            "200"
+        ],
+        "nginx.ingress_controller.upstream.response.time_total": 0.0,
+        "nginx.ingress_controller.upstream.response.time_list": [
+            "0.000"
+        ],
+        "nginx.ingress_controller.upstream_ip_list": [
```
Comment (jsoriano) [inline filebeat/module/nginx/ingress_controller/test/test.log-expected.json:44]: We are not storing ips in this field, maybe we can call it `upstream_endpoint_list`, or `upstream_address_list`.

File: filebeat/module/nginx/ingress_controller/test/test.log-expected.json [hunk 5]
```diff
@@ -1017,5 +1257,69 @@
...
+        "nginx.ingress_controller.upstream.alternative_name": "",
+        "nginx.ingress_controller.upstream.ip": "172.17.0.7",
+        "nginx.ingress_controller.upstream.name": "default-web2-8080",
+        "nginx.ingress_controller.upstream.port": 8080,
+        "nginx.ingress_controller.upstream.response.length_total": 161,
+        "nginx.ingress_controller.upstream.response.length_list": [
+            "61",
+            "100"
+        ],
+        "nginx.ingress_controller.upstream.response.status_code": 203,
```
Comment (jsoriano) [inline filebeat/module/nginx/ingress_controller/test/test.log-expected.json:1296]: Just to confirm, is this the status code received by the client? I think we should report the one received by the client here.

File: filebeat/module/nginx/ingress_controller/test/test.log-expected.json [hunk 6]
```diff
@@ -1017,5 +1257,69 @@
...
+        "fileset.name": "ingress_controller",
+        "http.request.method": "GET",
+        "http.response.body.bytes": 61,
+        "http.response.status_code": 200,
+        "http.version": "1.1",
+        "input.type": "log",
+        "log.offset": 5730,
+        "nginx.ingress_controller.http.request.id": "835136ae24486dbb4156dcbe21f5d402",
+        "nginx.ingress_controller.http.request.length": 348,
+        "nginx.ingress_controller.http.request.time": 0.001,
```
Comment (jsoriano) [inline filebeat/module/nginx/ingress_controller/test/test.log-expected.json:1283]: At some moment we should probably move these fields to ECS fields.

**Review by jsoriano — CHANGES_REQUESTED** [review]
Arrays are problematic, for example here we might want to use arrays of objects, so we have single objects for the address, status code and response type for each endpoint. But I think you have made a good compromise.

File: filebeat/module/nginx/ingress_controller/ingest/pipeline.yml [hunk 7]
```diff
@@ -41,6 +61,81 @@ processors:
...
+    if: ctx.nginx?.ingress_controller?.upstream?.response?.length_list != null && ctx.nginx.ingress_controller.upstream.response.length_list.length > 0
+    lang: painless
+    source: >-
+      try {
+        if (ctx.nginx.ingress_controller.upstream.response.length_list.length == null) {
+          return;
+        }
+        int length = 0;
+        for (def item : ctx.nginx.ingress_controller.upstream.response.length_list) {
+          length = length + Integer.parseInt(item);
```
Comment (chendo) [inline filebeat/module/nginx/ingress_controller/ingest/pipeline.yml:74]: Hmmm, does it make sense to sum up the response lengths? The client would only see the most recent response.

File: filebeat/module/nginx/ingress_controller/ingest/pipeline.yml (hunk 1 above)
Comment (ChrsMark) [reply filebeat/module/nginx/ingress_controller/ingest/pipeline.yml]: 👍

File: filebeat/module/nginx/ingress_controller/test/test.log-expected.json (hunk 5 above)
Comment (ChrsMark) [reply filebeat/module/nginx/ingress_controller/test/test.log-expected.json:1296]: Yes, the logic here is to report the last value here as this is the one the client receives.

File: filebeat/module/nginx/ingress_controller/test/test.log-expected.json (hunk 6 above)
Comment (ChrsMark) [reply filebeat/module/nginx/ingress_controller/test/test.log-expected.json:1283]: :) 👍

File: filebeat/module/nginx/ingress_controller/ingest/pipeline.yml (hunk 7 above)
Comment (ChrsMark) [reply filebeat/module/nginx/ingress_controller/ingest/pipeline.yml:74]: That makes sense, I will go with it! Thanks!

File: filebeat/module/nginx/ingress_controller/test/test.log-expected.json (hunk 5 above)
Comment (jsoriano) [reply filebeat/module/nginx/ingress_controller/test/test.log-expected.json:1296]: Perfect, thanks.

**Review by jsoriano — APPROVED** [review]

**Pull request merged**